### ISSUE TITLE
Question regarding the Request Context in the TLS 1.3 Certificate Server part

### ISSUE THREAD

**Issue by elecharny** [syncsynchalt/illustrated-tls13#2]
Hi!

First, thanks *a LOT* for this excellent peice of work !

I have a question on the **Request Context** byte in the **Certificate Request** message. You have it set to **0x00**, but when I decode some TLS 1.3 HandShake, I don't see this byte at all. Here is what I typically get:

```
Server certificate:
-------------------
0B 
00 03 85 	        certificate payload
00 03 82 		certificates length
00 03 7F 		certificate length
...

```

compared to what's in your document:

```
Server certificate:
-------------------
0b 
00 03 2e  	        certificate payload
00                           request context <<-------------
00 03 82 		certificates length
00 03 7F 		certificate length
...
```

In the RFC (https://datatracker.ietf.org/doc/html/rfc8446#section-4.3.2), it says:

_certificate_request_context: 
An opaque string which identifies the certificate request and which will be echoed in the client's Certificate message. The certificate_request_context MUST be unique within the scope of this connection (thus preventing replay of client CertificateVerify messages). **This field SHALL be zero length unless used for the post-handshake authentication exchanges described in Section 4.6.2.**_
`

Do you have any insight about that ?

Many thanks!

**Comment by syncsynchalt** [comment]
Hello!

First of all, a disclaimer: I made the current TLS 1.3 site _before_ support was officially added to any of the standard libraries (IIRC I did it using some unofficial patches to BoringSSL), so the exact messages might not match against e.g. what OpenSSL chooses to send.  It's on my todo list to re-do the TLS 1.3 site against a modern version of OpenSSL.  This disclaimer alone might explain why you see different messages, extensions, etc. being sent in my sample connection than in actual modern TLS libraries.  It's not that the connection I'm documenting is necessarily invalid or incompatible, just that it might make different decisions than various other implementations will.

Now with that out of the way, let's look at your question. \*reads\* Ok, I think I understand it. First, the issue isn't actually [section 4.3.2](https://datatracker.ietf.org/doc/html/rfc8446#section-4.3.2), that describes a message which doesn't appear in my sample connection (if present, it would be between "Server Encrypted Extensions" and "Server Certificate"). I think you understood this, and referenced it for context. Instead, I think the relevant section is [section 4.4.2](https://datatracker.ietf.org/doc/html/rfc8446#section-4.4.2), which describes the "Certificate" message (type 0x0B):
```
...
      struct {
          opaque certificate_request_context<0..2^8-1>;  <-- HERE
          CertificateEntry certificate_list<0..2^24-1>;
      } Certificate;
...
   certificate_request_context:  If this message is in response to a
      CertificateRequest, the value of certificate_request_context in
      that message.  Otherwise (in the case of server authentication),
      this field SHALL be zero length.
```

The question I have at this point is whether the statement "this field SHALL be zero length" means it should be zero bytes (not present), or whether it should be a single byte saying "variable length field of length zero". My site seems to think it's the second option, and Wireshark is agreeing, for what it's worth (not that that's proof; this could just be a common error that got grandfathered in). I think that [section 3.4 (Vectors)](https://datatracker.ietf.org/doc/html/rfc8446#section-3.4) is the relevant section for this:
```
Variable-length vectors are defined by specifying a subrange of legal
   lengths, inclusively, using the notation <floor..ceiling>.  When
   these are encoded, the actual length precedes the vector's contents
   in the byte stream.  The length will be in the form of a number
   consuming as many bytes as required to hold the vector's specified
   maximum (ceiling) length.  A variable-length vector with an actual
   length field of zero is referred to as an empty vector.

      T T'<floor..ceiling>;
```
I read this as an empty vector being the single byte 0x00, which matches what I'm seeing in my data capture.

Now, back to section 4.3.2, which I think is a red herring: it's an instruction for whether the _client_ should send its own authentication certificate to the _server_ in its authentication messages (in our case: no), but we're actually investigating the other direction in your question: the certificate that the _server_ sends to the _client_. So any data in a 4.3.2 message wouldn't apply to the bytes that you are looking at anyway, since they are from the server. This brings up an interesting point: nowhere in the protocol is a client allowed to ask a server for a particular certificate as far as I can tell, so the `certificate_request_context` field only seems to exist so that both the server->client and client->server messages can share a common format.  *This implies that in this particular message from the server to client, this field can only ever be 0x00!*  That sets off my radar, and now I'm considering three possibilities:

1. Is there an TLS extension in your connection but not in mine, which allows both client and server to agree that they shall leave out this field?
2. Is the TLS library used by your connections incorrect, and every other TLS implementation is working around the absent field since it can never be present anyway?
3. Or have I misread the RFC this morning, and the above is all wrong? 😉

If nothing else your question is making me want to re-do my page with OpenSSL 3.0 sooner than later; I have been worried that some strange difference might crop up and I would feel safer if I was documenting a much more widely deployed implementation.  I'll move that to the top of my todo list and we'll see what a new capture shows me.

**Comment by syncsynchalt** [comment]
Leaving this issue open until I re-do my site against OpenSSL 3.0+ (I'll keep the existing version available as well, for anyone currently puzzling through it). If nothing else I'll point to OpenSSL as a de-facto authority assuming it still has this extra field.

In the meantime let me know if you'd like to give me packet captures or let me know what TLS implementation you are testing against, this is an interesting difference that you're seeing.

**Comment by elecharny** [comment]
Hi ! Thanks for the clarification!

My bad, it was late and I misread the RFC. It's 4.4.2, absolutely (Certificate).

And my bad, 2: I was really convinced I was analyzing some TLS 1.3 handshake, but it was actually TLS 1.2... :/ (I'm using Java SslEngine, not OpenSSL, FTR)

Thanks a lot for the time you spent answering me !

**Comment by syncsynchalt** [comment]
That would explain it!  TLS 1.2 likely doesn't have the same challenge / response certificate token concept, and also won't have the unified certificate message format between the server / client directions which is what introduced this field.

It doesn't help that TLS 1.3 connections are marked as being version 1.2, for maddening reasons. 😄 

Thanks for the response, and you've still given me the energy to update my site.  Have a good day!

**Comment by elecharny** [comment]
Thanks !

To explain the issue, I was setting the server to TLs 1.3 but sadly, in an obscure area of the code, the client was using "TLS" (which resolves to TLS 1.1). Obviously, the protocol was downgraded on the server to 1.1, not 1.3, thus the difference I saw.

Anyway, your work, is really invaluable. The TLS 1.2/1.3 RFCs are really tedious to read, to say the least...

**Comment by syncsynchalt** [comment]
(For anyone following: https://tls13.ulfheim.net has been updated to OpenSSL 3.0.1, nice to finally have that off my todo list!)
